lib/Mojo/Promise.pm

```diff
@@ -20,14 +20,20 @@ sub all {
   my $results   = [];
   my $remaining = scalar @promises;
   for my $i (0 .. $#promises) {
-    $promises[$i]->then(
-      sub {
-        $results->[$i] = [@_];
-        $all->resolve(@$results) if --$remaining <= 0;
-      },
-      sub { $all->reject(@_) }
-    );
+    if ($promises[$i] && blessed $promises[$i] && $promises[$i]->can('then')) {
+      $promises[$i]->then(
+        sub {
+          $results->[$i] = [@_];
+          $all->resolve(@$results) if --$remaining <= 0;
+        },
+        sub { $all->reject(@_) }
+      );
+    } else {
+      $results->[$i] = [$promises[$i]];
+      --$remaining;
+    }
   }
+  $all->resolve(@$results) if $remaining <= 0; # all non-promises
```

Comment:
This would only work with the global `Mojo::IOLoop` singleton, since there is no chance to change the value of the `ioloop` attribute before a `next_tick` is initiated.

Reply:
And that means i believe for this pull request to be accepted we would have to deprecate the `ioloop` attribute first.